tests/test_service.py

```diff
@@ -0,0 +1,196 @@
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+from unittest.mock import Mock, patch
+
+import pytest
+from phabricatoremails import logging
+from phabricatoremails.db import DBNotInitializedError
+from phabricatoremails.mail import OutgoingEmail, FsMail
+from phabricatoremails.render.render import Render
+from phabricatoremails.render.template import TemplateStore
+from phabricatoremails.service import Pipeline, service
+from tests.mock_db import MockDB
+from tests.mock_mail import MockMail
+from tests.mock_settings import MockStats, MockSettings
+from tests.mock_source import MockSource
+from tests.mock_thread_store import MockThreadStore
+from tests.mock_worker import MockWorker
+
+
+def _assert_mail(mail: OutgoingEmail, subject, to, content_needle):
```

Comment:
Docstring missing and needed from this, and other similar methods in this and other modules.

Reply:
These are test methods that may change. I feel like the specifics of what are asserted here is best represented in the code itself.

Reply:
> What does "_assert_mail" mean? Either the name of this test should be improved, or you should add a docstring describing what this test is testing and how it can fail. There really isn't a hint as to what this method is trying to do in the name.

I can see one of two possible variant docstrings for this test:
1. A docstring that doesn't provide enough detail: "Asserts some properties on an outgoing email"
2. A docstring that parrots the implementation: "Asserts the subject, to properties of an outgoing email, and also confirms that a provided string exists in both the html and text contents"

Both of these solutions are worse than no docstring to me:
* One of them is a verbose string providing no detail - I already know it's asserting an email based on the name
* One of them is duplicating the implementation in text form, practically statement-by-statement. Every time the function changes (e.g.: a new property is asserted), the doc has to be updated, or it's going to be incorrect. The function is simple enough that reading the implementation should be sufficient.

> What is content_needle? It's not used or documented anywhere other than this method, and someone who is not familiar with an analogy of trying to find a needle in a haystack may not understand the purpose of this variable or why it is named so.

The [PHP standard library](https://www.php.net/manual/en/function.strpos.php) uses the "need and haystack" naming convention for some search functions. If it's not clear from the name, reading the implementation of `_assert_mail()` also makes the purpose of `content_needle` clear.

> If you don't want to include a docstring and put this in a docstring, then at least document it inline as "a unique string based on the specific type of email that is generated that should be present in the body of the email", or something similar.

This should be clear from the implementation and the way that the function is used.

> In this particular case, I don't particular see an advantage to factoring these asserts into their own method.

As I was writing the tests in `test_service`, I found that I was asserting each email in a consistent way - checking the subject, target email address, and lightly confirming the contents. This was repeated enough times that I realized the abstraction here - I was making assertions on a specific outgoing email. 

-----

Also, as reminder: this is a _private_ function used only within `test_service`, it doesn't need to be documented for external use.